src/Twig/Extension/PageExtension.php

```diff
@@ -86,48 +78,31 @@ class PageExtension extends AbstractExtension implements InitRuntimeInterface
      * @param BlockHelper                 $blockHelper        The Block Helper
      * @param bool                        $hideDisabledBlocks
      */
-    public function __construct(CmsManagerSelectorInterface $cmsManagerSelector, SiteSelectorInterface $siteSelector, RouterInterface $router, BlockHelper $blockHelper, HttpKernelExtension $httpKernelExtension, $hideDisabledBlocks = false)
+    public function __construct(CmsManagerSelectorInterface $cmsManagerSelector, SiteSelectorInterface $siteSelector, RouterInterface $router, BlockHelper $blockHelper, RequestStack $requestStack, $hideDisabledBlocks = false)
     {
         $this->cmsManagerSelector = $cmsManagerSelector;
         $this->siteSelector = $siteSelector;
         $this->router = $router;
         $this->blockHelper = $blockHelper;
-        $this->httpKernelExtension = $httpKernelExtension;
+        $this->requestStack = $requestStack;
         $this->hideDisabledBlocks = $hideDisabledBlocks;
     }
 
     public function getFunctions()
     {
         return [
             new TwigFunction('sonata_page_ajax_url', [$this, 'ajaxUrl']),
-            new TwigFunction('sonata_page_breadcrumb', [$this, 'breadcrumb'], ['is_safe' => ['html']]),
+            new TwigFunction('sonata_page_breadcrumb', [$this, 'breadcrumb'], ['is_safe' => ['html'], 'needs_environment' => true]),
             new TwigFunction('sonata_page_render_container', [$this, 'renderContainer'], ['is_safe' => ['html']]),
             new TwigFunction('sonata_page_render_block', [$this, 'renderBlock'], ['is_safe' => ['html']]),
             new TwigFunction('controller', [$this, 'controller']),
         ];
     }
 
-    public function initRuntime(Environment $environment): void
```

Comment:
@VincentLanglet just to clarify

in this case, ins't it a BC broken?

if someone is using this  `initRuntime` method, and the people didn't get any deprecation message, it's a BC broken, isn't it?

Reply:
that's why i did it for the 4.x branch so i can do BC breaks

Reply:
yes, But it wasn't deprecated in `3.x` then who uses `3.x` doesn't know that it was removed.

Reply:
ahh Ok ;)

Reply:
Also page Extension was "@final" so it shouldn't have been touched by users anyway